haskell.compiler.ghc96{3,4}: fix builds on aarch64-linux

sandydoo opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

~Relax stricter pointer checks for 9.6.0 - 9.6.4 on aarch64-linux.~
Patches 9.6.0 - 9.6.4 on aarch64-linux to fix an incompatible pointer type error when building the RTS.
https://gitlab.haskell.org/ghc/ghc/-/issues/24348

Fixed upstream starting with GHC 9.6.5.

Failing builds on Hydra: 
- https://hydra.nixos.org/job/nixpkgs/trunk/haskell.compiler.ghc964.aarch64-linux
- https://hydra.nixos.org/job/nixpkgs/trunk/haskell.compiler.ghc963.aarch64-linux

LMK if this is the right place/approach for this.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

maralorn commented:
Do we need this fix? Couldn’t we just drop those ghc versions?

sandydoo commented:
> Do we need this fix? Couldn’t we just drop those ghc versions?

Yeah, I suppose they'll be dropped in the next stable. haskell.nix currently uses one of these versions for bootstrapping and, while that will also eventually be updated, I figured it would be nice to have these versions build while they're available.

sandydoo commented:
Can we get this merged or should I close this?

Affected projects: https://github.com/input-output-hk/haskell.nix/issues/2337

maralorn commented:
I kinda would prefer to hear sternis opinion on this.

However I am confused by those haskell.nix references. Does haskell.nix use nixpkgs ghcs for bootstraping? Or are you just mentioning it because it has the same issues?

sandydoo commented:
> However I am confused by those haskell.nix references. Does haskell.nix use nixpkgs ghcs for bootstraping?

Correct. It will now try to use 9.6.6 if available: https://github.com/input-output-hk/haskell.nix/pull/2342/files

sternenseemann commented:
I think this is fine to fix, but I'd prefer to just pick the [upstream patch](https://gitlab.haskell.org/ghc/ghc/-/commit/1e48c43483693398001bfb0ae644a3558bf6a9f3).

sandydoo committed:
haskell.compiler.ghc96{3,4}: fix builds on aarch64-linux

Applies an upstream patch that fixes a pointer type mismatch in the RTS.
https://gitlab.haskell.org/ghc/ghc/-/issues/24348

Fixed upstream starting with GHC 9.6.5.

maralorn reviewed (approved):
Thank you!
